.circleci/config.yml

```diff
@@ -266,41 +267,74 @@ jobs:
           name: Upload Coverage
           command: bash <(curl -s https://codecov.io/bash) -cF python -s "/tmp/workspace/coverage/"
 
-  build_image:
+  build-docker-image:
     docker:
       - image: docker
+    parameters:
+      python-version:
+        type: string
+      tag-latest:
+        type: boolean
+        default: false
+    environment:
+      PYTHON_VERSION: << parameters.python-version >>
+      PYTHON_TAG: python<< parameters.python-version >>
     steps:
-      - setup_remote_docker
       - checkout
-      - run:
-          name: Docker Build
-          command: docker build -t prefecthq/prefect .
+      - setup_remote_docker:
+          docker_layer_caching: true
+      - run:
+          name: Build image
+          command: >-
+            docker build 
+            --build-arg GIT_SHA=$CIRCLE_SHA1 
+            --build-arg BUILD_DATE=$(date -u +'%Y-%m-%dT%H:%M:%SZ')
+            --build-arg PREFECT_VERSION=$CIRCLE_TAG 
+            -t prefecthq/prefect:${CIRCLE_TAG}-${PYTHON_TAG}
+            -t prefecthq/prefect:$PYTHON_TAG
+            .
+      - when:
+          condition: << parameters.tag-latest >>
+          steps:
+            - run:
+                name: Tag latest image
+                command: |
+                  docker tag prefecthq/prefect:${CIRCLE_TAG}-${PYTHON_TAG} prefecthq/prefect:latest
       - run:
           name: Test image
           command: |
             docker run -dit prefecthq/prefect /bin/bash -c 'curl -fL0 https://raw.githubusercontent.com/PrefectHQ/prefect/master/examples/retries_with_mapping.py | python'
       - run:
-          name: Authenticate with Docker Hub and push
+          name: Push versioned tags
           command: |
             docker login --username  $DOCKER_HUB_USER --password $DOCKER_HUB_PW
-            docker push prefecthq/prefect
+            docker push prefecthq/prefect:${CIRCLE_TAG}-${PYTHON_TAG}
```

Comment:
I think we might want to finesse the logic here a little, because I suspect this will end up pushing an image with a messy git tag

Reply:
I'm open to a better way to do it. I added [circle-ci git tag filters here](https://github.com/PrefectHQ/prefect/pull/1648/files/3aad2e7a686429f40856e074712a1dd4d3fc9d0e#diff-1d37e48f9ceff6d8030570cd36286a61R354-R355) to help reduce the risk of an odd tag coming through.

Reply:
Hmmm yea but won't this job still run on master?

Reply:
good call, I thought these were `and`ed instead of `or`ed. I'll update...

Reply:
Interestingly enough there isn't a good way to do this, @TylerWanner and I came up with this combo:
- https://github.com/PrefectHQ/prefect/pull/1648/files#diff-1d37e48f9ceff6d8030570cd36286a61R361-R364
- https://github.com/PrefectHQ/prefect/pull/1648/files#diff-1d37e48f9ceff6d8030570cd36286a61R286-R292

Up to a better approach, but this is all I've got until circleci supports and'ed filters.

Reply:
I think this is the only place I'm still hesitant; here's what I imagine happening:
- a PR is merged to master
- `build_docker_image` is run
- `CIRCLE_TAG` is empty, because this isn't a tagged commit
- either this step or the build step errors out in some unique way based on this

Everything else looks great though! And to be honest, we could always clean things up if that happens but I want to make sure I understand this piece

Reply:
I don't believe that it is possible to reach this step because of the workflow filter, however, I'm all for adding the precautionary `set -u` mentioned in the other comment below.